# A Beginner's Guide To Use Python For Linear Algebra

We are going to use Numpy library of Python. Numpy is the fundamental package for scientific computing in Python. It provides different data structures, functions and tools for numerical computing in Python. We are going to learn about these given below functions with working examples (useful in linear algebra):

- det
- matmul
- eigvals
- inv
- solve

The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.

In [1]:
!pip install jovian --upgrade -q

In [2]:
import jovian

In [3]:
jovian.commit(project='zerotoanalyst-numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Updating notebook "singhalkshama4343/zerotoanalyst-numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations


'https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations'

Let's begin by importing Numpy as `np` and listing out the functions covered in this notebook.

In [4]:
import numpy as np

To make it easy to call Numpy we are aliasing Numpy as np. 

`linalg` module contains all the necessary functions of linear algebra, so we have to call functions as `np.linalg.fun_name`.

## Function 1 - np.linalg.det
**Question**: What is area of the triangle having vertices `P(4,2), Q(8,4) and R(5,6)`? 

Area of the triangle having vertices P($x_1$,$y_1$), Q($x_2$,$y_2$) and R($x_3$,$y_3$) is given by
$$Area=\frac{1}{2}\begin{vmatrix}x_1 & y_1 & 1 \\
x_2 & y_2 & 1 \\
x_3 & y_3 & 1
\end{vmatrix}
$$

In [5]:
arr1 = [[4, 2,1], 
        [8, 4,1],
       [5,6,1]]
area=np.linalg.det(arr1)/2
area

6.999999999999998

Using the above formula we have calculated area of the triangle, here at the place of determinant we have used `np.linag.det` function of Numpy which find out the determinant of the matrix. Hence, the area is approximately $7$ $sq. units$.

**Question**: What is the determinant of the matrix $\begin{bmatrix}5 &6&7\\8&9&10\\3&5&97\end{bmatrix}$?

In [7]:
arr1 = [[5, 6, 7], 
        [8, 9, 10],
       [3,5,97]]
np.linalg.det(arr1)

-269.99999999999983

To find out the determinant of the above matrix we have used `np.linalg.det` function from Numpy library as it is same as determinant in Mathematics. Instead of doing it manually this function make it easy to find the determinant. Hence, the determinant is approximately $-270$.

**Question**: Can you find out the determinant of the matrix $\begin{bmatrix}5 &6&7\\8&9&10\end{bmatrix}$?

In [8]:
arr1 = [[5, 6, 7], 
        [8, 9, 10]]

np.linalg.det(arr1)

LinAlgError: Last 2 dimensions of the array must be square

This is breaking because it is possible to find out the determinant of a square matrix only. 

Use `np.linalg.det` to find out determinant of sqaure matrices only.

In [9]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "singhalkshama4343/zerotoanalyst-numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations


'https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations'

## Function 2 - np.linalg.matmul

**Question** What is matrix multiplication of $arr1=\begin{bmatrix}1&2\\3&4.0\end{bmatrix}$ and  $arr2=\begin{bmatrix}5&6&7\\8&9&10\end{bmatrix}$?

In [10]:
arr1 = [[1, 2], 
        [3, 4.0]]
arr2 = [[5, 6, 7], 
        [8, 9, 10]]
np.matmul(arr1,arr2)

array([[21., 24., 27.],
       [47., 54., 61.]])

`np.matmul` is used for matrix multiplication it is similar to matrix multiplication in linear algebra. The above code is giving numpy array as the output as in numpy array means N dimensional array while matrix is a 2 dimensional array. So, Matrix is a subset of numpy arrays.

**Question** What is matrix multiplication of $arr1=\begin{bmatrix}4&2\\8&12.0\end{bmatrix}$ and  $arr2=\begin{bmatrix}16&6\\9&10\end{bmatrix}$?

In [11]:
arr1 = [[4, 2], 
        [8, 12.0]]
arr2 = [[16, 6], 
        [9, 10]]
np.matmul(arr1,arr2)

array([[ 82.,  44.],
       [236., 168.]])

**Question** What is matrix multiplication of $arr1=\begin{bmatrix}5&6&7\\8&9&10\end{bmatrix}$ and  $arr2=\begin{bmatrix}1&2\\3&4.0\end{bmatrix}$?

In [12]:
arr1 = [[5, 6, 7], 
        [8, 9, 10]]
arr2 = [[1, 2], 
        [3, 4.0]]
np.matmul(arr1,arr2)

ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 2 is different from 3)

The above code is breaking as it is violating the matrix multiplication condition.

*Condition*: We can find out A$\times$B where matrix `A` have size $m\times n$ and `B` have size $x\times y$ only if $n=x$.

Keep in mind the above condition at the time of using `np.linalg.matmul` function.

In [13]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "singhalkshama4343/zerotoanalyst-numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations


'https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations'

## Function 3 - np.linalg.eigvals

**Question** What are the eigenvalues for the matrix $\begin{bmatrix}5&12\\26&4.0\end{bmatrix}$?

In [14]:
arr1 = [[5, 12], 
        [26, 4.0]]
np.linalg.eigvals(arr1)

array([ 22.17059705, -13.17059705])

**Question** What are the eigenvalues for the matrix $\begin{bmatrix}5&12&6\\26&4.0&7\\4&7&9\end{bmatrix}$?

In [15]:
arr1 = [[5, 12,6], 
        [26, 4.0,7],
       [4,7,9]]
np.linalg.eigvals(arr1)

array([-12.9913107 ,  26.5182498 ,   4.47306091])

To find out the eigenvalues for a given matrix use `np.linalg.eigvals` function. As the above is a $3\times3$ matrix so the output is an array containing $3$ eigenvalues.

**Question** What are the eigenvalues for the matrix $\begin{bmatrix}5&12&6\\26&4.0&7\end{bmatrix}$?

In [16]:
arr1 = [[5, 12,6], 
        [26, 4.0,7]]
np.linalg.eigvals(arr1)

LinAlgError: Last 2 dimensions of the array must be square

According to linear Algebra we can find out eigenvalues for square matrices only that's why the above code is breaking as the above matrix is not a square matrix.

Keep in mind that the matrix should be square at the time of using `np.linalg.eigvals`.

In [17]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "singhalkshama4343/zerotoanalyst-numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations


'https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations'

## Function 4 - np.linalg.inv

**Question** What will be the inverse for the matrix $\begin{bmatrix}5&12&6\\26&4.0&7\\4&7&9\end{bmatrix}$?

In [18]:
arr1 = [[5, 12,6], 
        [26, 4.0,7],
       [4,7,9]]
np.linalg.inv(arr1)

array([[ 0.00843608,  0.04282933, -0.03893576],
       [ 0.13367943, -0.01362751, -0.07852044],
       [-0.10772226, -0.00843608,  0.18948735]])

**Question** What will be the inverse for the matrix $\begin{bmatrix}1&2\\3&4.0\end{bmatrix}$?

In [19]:
arr1 = [[1, 2], 
        [3, 4.0]]
np.linalg.inv(arr1)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

We are using `np.linalg.inv` to find out the inverse for a matrix.

**Question** What will be the inverse for the matrix $\begin{bmatrix}1&2&7\\3&4.0&9\end{bmatrix}$?

In [20]:
arr1 = [[1, 2,7], 
        [3, 4.0,9]]
np.linalg.inv(arr1)

LinAlgError: Last 2 dimensions of the array must be square

According to linear Algebra we can find out inverse for square matrices only that's why the above code is breaking as the above matrix is not a square matrix.

Keep in mind that the matrix should be square at the time of using `np.linalg.inv`.

In [21]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "singhalkshama4343/zerotoanalyst-numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations


'https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations'

## Function 5 - np.linalg.solve

**Question** Solve the system of equations $x_0 + 2 x_1 = 1$ and $3 x_0 + 5 x_1 = 2$

In [22]:
a=[[1,2],
  [3,5]]
b=[[1],
   [2]]
np.linalg.solve(a,b)

array([[-1.],
       [ 1.]])

To solve the linear system of equation we have used `np.linalg.solve` function. Hence, the solution of $x_0$ is $-1$ and $x_1$ is $1$.

**Question** Solve the system of equations $x_0 + 2 x_1 + 3 x_2 = 4 $ and $3 x_0 + 5 x_1 + 7 x_2= 8$ and $4 x_0+ 9 x_1+ 8 x_2=10$

In [23]:
a=[[1,2,3],
  [3,5,7],
  [4,9,8]]
b=[[4],
   [8],
  [10]]
np.linalg.solve(a,b)

array([[-2.33333333],
       [ 0.66666667],
       [ 1.66666667]])

Hence, the solution of $x_0$ is $-2.33$, $x_1$ is $0.67$ and $x_3$ is $1.67$ after rounding off the solutions to 2 digits.

**Question** Solve the system of equations $x_0 + 2 x_1 + 3 x_2 + 4 x_3 = 4 $ and $3 x_0 + 5 x_1 + 7 x_2 + 9 x_3= 8$ and $4 x_0+ 9 x_1+ 8 x_2 +8 x_3=10$

In [24]:
a=[[1,2,3,4],
  [3,5,7,9],
  [4,9,8,8]]
b=[[4],
   [8],
  [10]]
np.linalg.solve(a,b)

LinAlgError: Last 2 dimensions of the array must be square

The above code is breaking because `a` matrix is not a square matrix.

`a` must be square and of full-rank, i.e., all rows (or, equivalently, columns) must be linearly independent; if either is not true, use `lstsq` for the least-squares best “solution” of the system/equation.

In [25]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "singhalkshama4343/zerotoanalyst-numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations


'https://jovian.ai/singhalkshama4343/zerotoanalyst-numpy-array-operations'

## Conclusion
 We have learnt about these given below functions with working examples (useful in linear algebra):

- det
- matmul
- eigvals
- inv
- solve

These were basic functions helpful in linear algebra. If you want to go advance, there are some functions which will be good to know:

- svd (use for singular value decomposition)
- eigval (use to find out eigenvalues and eigenvectors)
- norm (use to find norm) 
and many more. 

Check out [this](https://numpy.org/doc/stable/reference/routines.linalg.html) to know more functions.


## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* Numpy Linear Algebra official: https://numpy.org/doc/stable/reference/routines.linalg.html 

In [ ]:
jovian.commit()

<IPython.core.display.Javascript object>